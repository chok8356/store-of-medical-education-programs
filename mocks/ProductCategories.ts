import { ICategory } from '../types/ICategory'

export const PRODUCTS_CATEGORIES_COUNT = 5

export const categories: ICategory[] = new Array(PRODUCTS_CATEGORIES_COUNT)
  .fill(null)
  .map((x, i) => {
    const id = i + 1
    return {
      id,
      name: `Категория ${id}`
    }
  })
