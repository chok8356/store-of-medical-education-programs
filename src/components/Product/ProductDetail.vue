<template>
  <b-card
    v-if="product"
    class="overflow-hidden">
    <b-row>
      <b-col md="6">
        <b-card-img :src="product.img" />
      </b-col>
      <b-col md="6">
        <b-card-body>
          <h1 class="mb-4">
            {{ product.name }}
          </h1>

          <b-card-text>
            {{ product.description }}
          </b-card-text>

          <b-badge variant="info">
            {{ category.name }}
          </b-badge>

          <div class="d-block mt-4">
            <b-button
              variant="primary"
              @click="addToCart">
              В корзину
            </b-button>
          </div>
        </b-card-body>
      </b-col>
    </b-row>
  </b-card>
</template>

<script lang="ts">
import { BCard, BRow, BCol, BCardBody, BCardText, BBadge, BButton, BCardImg } from 'bootstrap-vue'
import { PropType } from 'vue'
import { Vue, Component, Prop } from 'vue-property-decorator'
import { ICategory } from '../../../types/ICategory'
import { IProduct } from '../../../types/IProduct'

@Component({
  name: 'ProductDetail',
  components: {
    BCard, BRow, BCol, BCardBody, BCardText, BBadge, BButton, BCardImg
  }
})
export default class ProductDetail extends Vue {
  @Prop({ type: Object as PropType<IProduct> }) product!: IProduct
  @Prop({ type: Object as PropType<ICategory> }) category!: ICategory

  addToCart() {
    this.$emit('add-to-cart')
  }
}
</script>
