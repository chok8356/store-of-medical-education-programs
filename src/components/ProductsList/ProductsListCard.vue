<template>
  <b-card
    class="products-list-card"
    :img-src="product.img">
    <b-card-text class="products-list-card__title mb-2">
      {{ product.name }}
    </b-card-text>

    <b-card-text class="font-weight-bold mb-2">
      {{ product.price | currency }}
    </b-card-text>

    <b-card-text class="products-list-card__description">
      {{ product.description | truncate(54) }}
    </b-card-text>

    <b-button-group size="sm">
      <b-button
        variant="outline-primary"
        @click="addToCart">
        В корзину
      </b-button>
      <b-button
        variant="outline-secondary"
        @click="showDetail">
        Подробнее
      </b-button>
    </b-button-group>
  </b-card>
</template>

<script lang="ts">
import { BCard, BCardText, BButtonGroup, BButton } from 'bootstrap-vue'
import { PropType } from 'vue'
import { Vue, Component, Prop } from 'vue-property-decorator'
import { truncate, currency } from '@/filters'
import { IProduct } from '../../../types/IProduct'

@Component({
  name: 'ProductsListCard',
  filters: {
    truncate, currency
  },
  components: {
    BCard, BCardText, BButtonGroup, BButton
  }
})
export default class ProductsListCard extends Vue {
  @Prop({ type: Object as PropType<IProduct> }) product!: IProduct

  addToCart() {
    this.$emit('add-to-cart')
  }

  showDetail() {
    this.$emit('show-detail')
  }
}
</script>

<style lang="scss">
.products-list-card {
  cursor: pointer;

  &__title {
    font-size: 16px;
    font-weight: 500;
  }

  &__description {
    font-size: 13px;
  }
}
</style>
