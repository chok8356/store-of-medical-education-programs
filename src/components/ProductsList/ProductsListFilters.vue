<template>
  <div
    v-if="filters"
    class="products-list-filters card p-4 shadow-sm">
    <b-form-group
      class="products-list-filters__search"
      label="Программы НМО">
      <b-form-input
        placeholder="Поиск"
        type="text"
        :value="filters.query"
        @input="onQueryInput" />
    </b-form-group>

    <hr class="mt-3 mb-4">

    <div class="products-list-filters__categories">
      <h5 class="d-block mb-3">
        Категории
      </h5>
      <b-nav
        pills
        vertical>
        <b-nav-item
          :active="filters.selectedCategory === null"
          @click="onCategoryClick(null)">
          Все категории
        </b-nav-item>
        <b-nav-item
          v-for="category in categories"
          :key="category.id"
          :active="filters.selectedCategory === category.id"
          @click="onCategoryClick(category.id)">
          {{ category.name }}
        </b-nav-item>
      </b-nav>
    </div>
  </div>
</template>

<script lang="ts">
import { BNav, BNavItem, BFormGroup, BFormInput } from 'bootstrap-vue'
import { PropType } from 'vue'
import { Vue, Component, Prop, Model } from 'vue-property-decorator'
import { ICategory } from '../../../types/ICategory'
import { IProductFilters } from '../../../types/IProduct'

@Component({
  name: 'ProductsListFilters',
  components: {
    BNav, BNavItem, BFormGroup, BFormInput
  }
})
export default class ProductsListFilters extends Vue {
  @Model('change', { type: Object as PropType<IProductFilters> }) filters!: IProductFilters
  @Prop({ type: Array as PropType<ICategory[]> }) categories!: ICategory[]

  onCategoryClick(id: ICategory['id'] | null) {
    this.$emit('change', {
      ...this.filters,
      selectedCategory: id
    })
  }

  onQueryInput(value: string) {
    this.$emit('change', {
      ...this.filters,
      query: value
    })
  }
}
</script>

<style lang="scss">
.products-list-filters {
  position: sticky;
  top: 104px;
  z-index: 1;
}
</style>
