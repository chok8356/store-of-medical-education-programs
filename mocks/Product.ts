import { IProduct } from '../types/IProduct'
import { PRODUCTS_CATEGORIES_COUNT } from './ProductCategories'

const PRODUCTS_COUNT = 100

export const products: IProduct[] = new Array(PRODUCTS_COUNT).fill(null).map((x, i) => {
  const id = i + 1
  return {
    id,
    name: `Продукт ${id}`,
    price: id + PRODUCTS_COUNT,
    // eslint-disable-next-line max-len
    img: 'https://media.istockphoto.com/photos/stack-of-medical-student-textbooks-picture-id921644532?k=20&m=921644532&s=612x612&w=0&h=hezQpZ0uyUmbU4QJePKK5JPNWVXD_UWwWPYSg5Rspds=',
    // eslint-disable-next-line max-len
    description: id % 7 + ' Полностью новые главы по биохимии и клеточной биологии, генетике, нервной системе, костям, мышцам и коже, эндокринной и репродуктивной системам, сердечно-сосудистой системе, почечной системе, диете и питанию.',
    category: id % PRODUCTS_CATEGORIES_COUNT + 1
  }
})
