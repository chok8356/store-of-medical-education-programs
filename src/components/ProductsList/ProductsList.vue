<template>
  <div class="products-list">
    <div class="row m-n2">
      <template v-if="products.length">
        <div
          v-for="product in products"
          :key="product.id"
          class="products-list__item col-md-4 p-2">
          <ProductsListCard
            :product="product"
            @add-to-cart="addToCart(product)"
            @show-detail="showDetail(product)" />
        </div>
      </template>
      <div
        v-else
        class="d-block text-center text-muted p-5">
        Нет продуктов.
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { Vue, Component, Prop } from 'vue-property-decorator'
import ProductsListCard from '@/components/ProductsList/ProductsListCard.vue'
import { IProduct } from '../../../types/IProduct'

@Component({
  name: 'ProductsList',
  components: { ProductsListCard }
})
export default class ProductsList extends Vue {
  @Prop({ type: Array, default: () => [] }) products!: IProduct[]

  addToCart(product: IProduct) {
    this.$emit('add-to-cart', product)
  }

  showDetail(product: IProduct) {
    this.$emit('show-detail', product)
  }
}
</script>

<style lang="scss">
.products-list {
  display: block;
  width: 100%;
}
</style>
