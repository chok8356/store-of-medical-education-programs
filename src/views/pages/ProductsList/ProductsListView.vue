<template>
  <div class="products-list-view">
    <div class="row">
      <div class="col-3">
        <ProductsListFilters
          v-loading="isLoadingCategories"
          :categories="categories"
          :filters="filters"
          @change="onFiltersChange" />
      </div>
      <div class="col-9">
        <ProductsList
          v-loading="isLoadingProducts"
          :products="filteredProducts"
          @add-to-cart="onAddToCart"
          @show-detail="onShowDetail" />
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { Vue, Component } from 'vue-property-decorator'
import { namespace } from 'vuex-class'
import ProductsList from '@/components/ProductsList/ProductsList.vue'
import ProductsListFilters from '@/components/ProductsList/ProductsListFilters.vue'
import { loading } from '@/directives'
import { ICategory } from '../../../../types/ICategory'
import { IProduct, IProductFilters } from '../../../../types/IProduct'

const ProductStore = namespace('product')
const ProductFiltersStore = namespace('productFilters')
const CategoryStore = namespace('category')
const CartStore = namespace('cart')
@Component({
  name: 'ProductsListView',
  components: { ProductsList, ProductsListFilters },
  directives: {
    loading
  }
})
export default class ProductsListView extends Vue {
  @ProductStore.Getter filteredProducts: IProduct[]
  @ProductStore.Action getProducts: any

  @ProductFiltersStore.State filters: IProductFilters
  @ProductFiltersStore.Mutation('SET_FILTERS') setFilters: any

  @CategoryStore.State categories: ICategory[]
  @CategoryStore.Action getCategories: any

  @CartStore.Action addToCart: any

  isLoadingCategories = true
  isLoadingProducts = true

  created() {
    this.getProducts().then(() => {
      this.isLoadingProducts = false
    })

    this.getCategories().then(() => {
      this.isLoadingCategories = false
    })
  }

  onFiltersChange(value: IProductFilters) {
    this.setFilters(value)
  }

  onAddToCart(product: IProduct) {
    this.addToCart(product.id)
  }

  onShowDetail(product: IProduct) {
    this.$router.push({ name: 'ProductDetail', params: { id: product.id.toString() } })
  }
}
</script>
