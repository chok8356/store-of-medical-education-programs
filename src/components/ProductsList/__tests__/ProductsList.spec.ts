import { shallowMount } from '@vue/test-utils'

import { products } from '@/../mocks/Product'
import ProductsListCard from '@/components/ProductsList/ProductsListCard.vue'
import ProductsList from '../ProductsList.vue'

const productsData = products.slice(0, 3)

describe('ProductsList', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(ProductsList, {
      propsData: {
        products: productsData
      }
    })
  })

  afterEach(() => {
    wrapper.destroy()
  })

  it('render correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('exist props', () => {
    expect(wrapper.props('products')).toStrictEqual(productsData)
  })

  it('emit add to cart', () => {
    const productsListCardCmp = wrapper.findComponent(ProductsListCard)
    productsListCardCmp.vm.$emit('add-to-cart')
    expect(wrapper.emitted('add-to-cart')[0][0]).toStrictEqual(productsData[0])
  })

  it('emit show detail', () => {
    const productsListCardCmp = wrapper.findComponent(ProductsListCard)
    productsListCardCmp.vm.$emit('show-detail')
    expect(wrapper.emitted('show-detail')[0][0]).toStrictEqual(productsData[0])
  })
})
