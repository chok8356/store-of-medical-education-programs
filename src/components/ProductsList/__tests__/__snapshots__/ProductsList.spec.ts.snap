// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsList render correctly 1`] = `
<div
  class="products-list"
>
  <div
    class="row m-n2"
  >
    <div
      class="products-list__item col-md-4 p-2"
    >
      <productslistcard-stub
        product="[object Object]"
      />
    </div>
    <div
      class="products-list__item col-md-4 p-2"
    >
      <productslistcard-stub
        product="[object Object]"
      />
    </div>
    <div
      class="products-list__item col-md-4 p-2"
    >
      <productslistcard-stub
        product="[object Object]"
      />
    </div>
  </div>
</div>
`;
